dandiapi/api/models/asset.py

```diff
@@ -117,6 +132,34 @@ def save(self, *args, **kwargs):
         self._populate_metadata()
         super().save(*args, **kwargs)
 
+    def generate_metadata(self, version):
+        """
+        Generate the correct metadata for this asset given a version.
+
+        The route used to download an asset will be different depending on which version it belongs
+        to. Since the download url is included in the asset metadata, it must be injected rather
+        than stored. This method should always be used instead of asset.metadata.metadata.
+        """
+        # TODO use http://localhost:8000 for local deployments
+        download_url = 'https://api.dandiarchive.org' + reverse(
+            'asset-download',
+            kwargs={
+                'versions__dandiset__pk': version.dandiset.identifier,
+                'versions__version': version.version,
+                'asset_id': str(self.asset_id),
+            },
+        )
+
+        blob_url = self.blob.blob.url
+        metadata = {
+            **self.metadata.metadata,
+            'identifier': str(self.asset_id),
+            'contentUrl': [download_url, blob_url],
+            'contentSize': self.blob.size,
+            'digest': self.blob.digest,
+        }
```

Comment:
a few comments on this section. 

1. we will likely add a few things here similar to dandiset metadata (`id`, `context`, `schemaKey`)
1. is self.blob.blob.url now devoid of a signed url? it can be a versioned url, but we want to drop signed urls from the metadata return and metadata export on publish. the signed urls should only be used in the context of the UI, because content-disposition needs to be added, unless we can figure a different way out to do so.

Reply:
outside the PR is fine, but that should be done before this PR is merged. 

the blob url bit is important since those presigned urls expire and would definitely not be useful as a publish link. if we are validating the info should be valid/proper.